[bitnami/grafana-operator] Added grafana-operator Chart

Cellebyte opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

**Description of the change**

Add a Helm Chart to install the grafana-operator to a Kubernetes Cluster.
Currently it does not provide a default Deployment for a Grafana installation.

<!-- Describe the scope of your change - i.e. what the change does. -->

**Benefits**
It solves the current problem of how to manage dashboards for applications.

<!-- What benefits will be realized by the code change? -->

**Possible drawbacks**

<!-- Describe any known limitations with your change -->

**Applicable issues**

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
  - fixes #[https://github.com/integr8ly/grafana-operator/issues/151](https://github.com/integr8ly/grafana-operator/issues/151)
  - fixes partially #2797
  - succeeds [https://github.com/integr8ly/grafana-operator/pull/153](https://github.com/integr8ly/grafana-operator/pull/153)

**Additional information**

  - It currently lacks support for helm 2
<!-- If there's anything else that's important and relevant to your pull
request, mention that information here.-->

**Checklist** <!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [X] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

joancafom commented on bitnami/grafana-operator/Chart.yaml:
```diff
@@ -0,0 +1,40 @@
+apiVersion: v2
```
We should be using `v1` because support for Helm v2 is also provided

Cellebyte replied:
Yeah I need to work on this ;)

joancafom commented on bitnami/grafana-operator/Chart.yaml:
```diff
@@ -0,0 +1,40 @@
+apiVersion: v2
+name: grafana-operator
+description: A Helm chart for Kubernetes
```
This can be replaced by something more descriptive, e.g:

"Kubernetes Operator based on the Operator SDK for creating and managing Grafana instances"

What do you think?

joancafom commented on bitnami/grafana-operator/values.yaml:
```diff
@@ -0,0 +1,154 @@
+## ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
+##
+podAnnotations: {}
+
```
It would be a nice idea to include `commonLabels` and `commonAnnotations`, that will be rendered for all deployed resources:

https://github.com/bitnami/charts/blob/5a5807c1b1db1f2337f6aa5308d3ff73a4329e6a/bitnami/ejbca/values.yaml#L142-L148

Cellebyte replied:
Already done.

1337andre commented on bitnami/grafana-operator/templates/deployment.yaml:
```diff
@@ -0,0 +1,106 @@
+          - -grafana-image={{ include "grafana-operator.grafana.baseImage" . }}
+          - -grafana-image-tag={{ .Values.grafana.image.tag | quote}}
+          - -grafana-plugins-init-container-image={{ include "grafana-operator.pluginInit.baseImage" . }}
+          - -grafana-plugins-init-container-tag={{ .Values.grafanaPluginInit.image.tag | quote}}
```
Failed to apply default image tag "quay.io/integreatly/grafana_plugins_init:\"0.0.3\"": couldn't parse image reference "quay.io/integreatly/grafana_plugins_init:\"0.0.3\"": invalid reference format

tested with k8s v.1.17.5

Cellebyte replied:
@1337andre I fixed the templating bug.
Could you please try again?

Cellebyte commented:
Good Morning Stale Bot,

It is not stale it only needs some further development.

beltran-rubo commented:
You are right @Cellebyte, added the proper tag to skip the stale bot. We plan to work on this new PR in the 3-4 next weeks.

Cellebyte commented:
@joancafom I now hopefully addressed all your feedback and improvements.
I will look into improving my metallb chart as well.
Thanks for the great common library.

andresbono commented:
Thanks so much for your valuable contributions, @Cellebyte!! ❤️ 

As @joancafom said in https://github.com/bitnami/charts/pull/3287#pullrequestreview-460722577, we need to work on the container images and we will give another review to the PR.

andresbono commented:
Hi @Cellebyte, just to let you know, it was not possible for us to work on the containers images needed for this new chart and it is very likely that it is not going to be possible at least in the following next two weeks. Sorry for that. We will keep this PR updated.

Cellebyte commented:
@andresbono I don't know if this chart should also provision a Grafana Instance.

andresbono commented:
> @andresbono I don't know if this chart should also provision a Grafana Instance.

Hi @Cellebyte, probably yes, It would be good to have a Grafana manifest like this one, right? https://github.com/integr8ly/grafana-operator/blob/14a623c4b4468f5321c2a75004434196bf4d19ff/deploy/examples/Grafana.yaml#L2

andresbono commented:
Hi @Cellebyte, we finally released the new container image for Grafana Operator. You can find the first (and latest) release here:

- `bitnami/grafana-operator` [`3.6.0-debian-10-r0`](https://github.com/bitnami/bitnami-docker-grafana-operator/commit/d9952f564cf4129bd9490ee965168bfccc1a21a0)

In addition to that, we have made some modifications in the `bitnami/grafana` container to make it fully compatible with the operator. We have basically doing 2 changes. We create a set of symbolic links to use the directory paths expected by the operator. We adapted the functions to download the plugins so the `bitnami/grafana` container can act as plugin init container. This is the release that introduced the changes:

- `bitnami/grafana` [`7.3.3-debian-10-r1`](https://github.com/bitnami/bitnami-docker-grafana/commit/ca9bb328fe0dc672906e035c089d85f514e3298e)

---

Could you please give them a try to check how the chart behaves? We have done some testing on our side, but it would be great if you can test it as well and we can continue working to have this PR merged.

Here you have the things we have tested:

<details><summary><b>Testing done (click me)</b></summary>

```console
$ kubectl create namespace grafana

$ helm install --namespace grafana gob . \
    --set image.registry=docker.io,image.repository=bitnami/grafana-operator,image.tag=3.6.0-debian-10-r0 \
    --set grafana.image.tag=7.3.3-debian-10-r1 \
    --set grafanaPluginInit.image.registry=docker.io,grafanaPluginInit.image.repository=bitnami/grafana,grafanaPluginInit.image.tag=7.3.3-debian-10-r1

# Watch operator logs
$ kubectl -n grafana logs gob-grafana-operator-5bc88fdfcf-28g8g -f

# Create a grafana instance
$ cat <<EOF | kubectl -n grafana apply -f -
apiVersion: integreatly.org/v1alpha1
kind: Grafana
metadata:
  name: example-grafana
spec:
  ingress:
    enabled: False
  config:
    log:
      mode: "console"
      level: "warn"
    auth:
      disable_login_form: False
      disable_signout_menu: True
    auth.anonymous:
      enabled: True
  dashboardLabelSelector:
    - matchExpressions:
        - { key: app, operator: In, values: [grafana] }
EOF

# Watch grafana until ready
$ kubectl -n grafana get pod grafana-deployment-7bb56dcc4b-rzg8h -w

# Create dashboard and dashboard with jsonnet
$ cat <<EOF | kubectl -n grafana apply -f -
apiVersion: integreatly.org/v1alpha1
kind: GrafanaDashboard
metadata:
  name: grafonnet-dashboard
  labels:
    app: grafana
spec:
  name: grafonnet-dashboard.json
  jsonnet: |-
    local grafana = import 'grafonnet/grafana.libsonnet';

    grafana + {
      "title": "Grafonnet Dashboard with jsonnet",
    } + {
      "id": null,
      "tags": [],
      "style": "dark",
      "timezone": "browser",
      "editable": true,
      "hideControls": false,
      "graphTooltip": 1,
      "panels": [],
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {
        "time_options": [],
        "refresh_intervals": []
      },
      "templating": {
        "list": []
      },
      "annotations": {
        "list": []
      },
      "refresh": "5s",
      "schemaVersion": 17,
      "version": 0,
      "links": []
    }
---
# https://github.com/integr8ly/grafana-operator/blob/master/deploy/examples/dashboards/SimpleDashboard.yaml
apiVersion: integreatly.org/v1alpha1
kind: GrafanaDashboard
metadata:
  name: simple-dashboard
  labels:
    app: grafana
spec:
  name: simple-dashboard.json
  json: >
    {
      "id": null,
      "title": "Simple Dashboard",
      "tags": [],
      "style": "dark",
      "timezone": "browser",
      "editable": true,
      "hideControls": false,
      "graphTooltip": 1,
      "panels": [],
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {
        "time_options": [],
        "refresh_intervals": []
      },
      "templating": {
        "list": []
      },
      "annotations": {
        "list": []
      },
      "refresh": "5s",
      "schemaVersion": 17,
      "version": 0,
      "links": []
    }
EOF

# Create dashboard with plugins
$ cat <<EOF | kubectl -n grafana apply -f -
# https://github.com/integr8ly/grafana-operator/blob/master/deploy/examples/dashboards/DashboardWithPlugins.yaml
apiVersion: integreatly.org/v1alpha1
kind: GrafanaDashboard
metadata:
  name: dashboard-with-plugins
  labels:
    app: grafana
spec:
  name: dashboard-with-plugins.json
  json: >
    {
      "id": null,
      "title": "Simple Dashboard with Plugins",
      "tags": [],
      "style": "dark",
      "timezone": "browser",
      "editable": true,
      "hideControls": false,
      "graphTooltip": 1,
      "panels": [],
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {
        "time_options": [],
        "refresh_intervals": []
      },
      "templating": {
        "list": []
      },
      "annotations": {
        "list": []
      },
      "refresh": "5s",
      "schemaVersion": 17,
      "version": 0,
      "links": []
    }
  plugins:
    - name: "grafana-piechart-panel"
      version: "1.5.0"
    - name: "grafana-clock-panel"
      version: "1.1.0"
EOF

# Check plugins
$ kubectl -n grafana exec -it grafana-deployment-859f99fb8d-7kkcn -- grafana-cli plugins ls

# Update plugin
$ cat <<EOF | kubectl -n grafana apply -f -
# https://github.com/integr8ly/grafana-operator/blob/master/deploy/examples/dashboards/DashboardWithPlugins.yaml
apiVersion: integreatly.org/v1alpha1
kind: GrafanaDashboard
metadata:
  name: dashboard-with-plugins
  labels:
    app: grafana
spec:
  name: dashboard-with-plugins.json
  json: >
    {
      "id": null,
      "title": "Simple Dashboard with Plugins",
      "tags": [],
      "style": "dark",
      "timezone": "browser",
      "editable": true,
      "hideControls": false,
      "graphTooltip": 1,
      "panels": [],
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {
        "time_options": [],
        "refresh_intervals": []
      },
      "templating": {
        "list": []
      },
      "annotations": {
        "list": []
      },
      "refresh": "5s",
      "schemaVersion": 17,
      "version": 0,
      "links": []
    }
  plugins:
    - name: "grafana-piechart-panel"
      version: "1.5.0"
    - name: "grafana-clock-panel"
      version: "1.1.1"
EOF

# Check plugins
$ kubectl -n grafana exec -it grafana-deployment-859f99fb8d-7kkcn -- grafana-cli plugins ls

# Get credentials
$ kubectl -n grafana exec grafana-deployment-859f99fb8d-7kkcn -- env | grep GF_
$ kubectl port-forward --namespace grafana svc/grafana-service 3000:3000

# Go to settings: http://localhost:3000/admin/settings
# Go to dashboards: http://localhost:3000/dashboards?layout=list
```

</details>

Keep us in the loop!

Cellebyte commented:
@andresbono I worked arround this issues with a configmap and a secret in the grafana instance configuration for the operator.
I will add a basic instance to the chart as a dedicated point in the chart. Where you can defined configuration for the grafana.
I could also think of a separate Chart. So that you can start deploying multiple Grafana Instances with it and referencing this chart here.

javsalgar commented:
Hi! Thanks for the input! I will continue checking the PR so please ping me when the planned changes you mentioned are ready :D

carrodher commented on bitnami/grafana-operator/Chart.yaml:
```diff
@@ -0,0 +1,31 @@
+apiVersion: v2
```
We're also standardizing the _Chart.yaml_ format by adding here the dependencies (already done) and also sorting alphabetically the fields and prettifying the format, you can do it automatically by running:
```
$ yq r --tojson Chart.yaml | jq . | yq r --prettyPrint - > Chart.yaml.new
$ mv Chart.yaml.new Chart.yaml
```

carrodher commented on bitnami/grafana-operator/Chart.yaml:
```diff
@@ -0,0 +1,31 @@
+
+dependencies:
+  - name: common
+    version: 0.x.x
```
There is a new major version for the common subchart
```suggestion
    version: 1.x.x
```

carrodher commented on bitnami/grafana-operator/README.md:
```diff
@@ -0,0 +1,136 @@
+## Prerequisites
+
+- Kubernetes 1.12+
+- Helm 2.12+ or Helm 3.0-beta3+
```
Helm 2 is not supported anymore when using `apiVersion: v2` in the _Chart.yaml_
```suggestion
- Helm 3.0-beta3+
```

carrodher commented on bitnami/grafana-operator/values.yaml:
```diff
@@ -0,0 +1,211 @@
+updateStrategy:
+  type: RollingUpdate
+
+image:
+  registry: quay.io
+  repository: integreatly/grafana-operator
+  tag: v3.6.0
```
```suggestion
image:
  registry: docker.io
  repository: bitnami/grafana-operator
  tag: 3.6.0-debian-10-r4
```

carrodher commented on bitnami/grafana-operator/values.yaml:
```diff
@@ -0,0 +1,211 @@
+    # pullSecrets:
+    #   - myRegistryKeySecretName
+
+grafanaPluginInit:
```
In this case, we need to find an alternative since this image is not part of the bitnami catalog

andresbono replied:
We adapted the functions to download the plugins so the `bitnami/grafana` container can act as plugin init container. This is the release that introduced the changes:

- `bitnami/grafana` [`7.3.3-debian-10-r1`](https://github.com/bitnami/bitnami-docker-grafana/commit/ca9bb328fe0dc672906e035c089d85f514e3298e)

carrodher commented:
Thanks for the changes! I just implemented a couple of minor modifications. Now the pending thing we need to do is related to the grafanaPluginInit image as it is not currently in the catalog

andresbono commented on bitnami/grafana-operator/values.yaml:
```diff
@@ -0,0 +1,205 @@
+
+grafanaPluginInit:
+  image:
+    registry: quay.io
+    repository: integreatly/grafana_plugins_init
+    tag: 0.0.3
```
```suggestion
    registry: docker.io
    repository: bitnami/grafana
    tag: 7.3.4-debian-10-r1
```

andresbono reviewed (commented):
Another batch of suggestions!

andresbono commented:
Hi @Cellebyte, thank you very much for addressing the suggestions. Right now, I think the only pending things are related to this comment: 

https://github.com/bitnami/charts/pull/3287#issuecomment-731567848

> I will add a basic instance to the chart as a dedicated point in the chart. Where you can define configuration for the Grafana.
>
> I could also think of a separate Chart. So that you can start deploying multiple Grafana Instances with it and referencing this chart here.

Do you think it makes sense to merge the PR as it is right now and leave those improvements that you suggest for future PRs?

Cellebyte commented:
> Hi @Cellebyte, thank you very much for addressing the suggestions. Right now, I think the only pending things are related to this comment:
> 
> [#3287 (comment)](https://github.com/bitnami/charts/pull/3287#issuecomment-731567848)
> 
> > I will add a basic instance to the chart as a dedicated point in the chart. Where you can define configuration for the Grafana.
> > I could also think of a separate Chart. So that you can start deploying multiple Grafana Instances with it and referencing this chart here.
> 
> Do you think it makes sense to merge the PR as it is right now and leave those improvements that you suggest for future PRs?

No I want to have also a basic grafana instance in it.

andresbono commented:
Thanks, @Cellebyte! 👍  We will keep this one still open in that case.

Cellebyte commented:
@joancafom @andresbono @carrodher: I updated the chart and the documentation I wait for a new review :+1: 
Thanks for your time revisiting the code I contributed ;)
Hopefully it's now smoot enough to get to upstream :)

Cellebyte commented:
@andresbono when running this chart with a securityContext it does not work.
the plugins directory does not have the correct uid and gid set

```console
1001@5677fdebe960:/opt/bitnami/grafana/data$ ls -lah
> total 12K
> drwxrwxr-x  3 root root 4.0K Dec 20 20:57 .
> drwxr-xr-x 11 1001 1001 4.0K Dec 20 20:58 ..
> drwxrwxr-x  2 root root 4.0K Dec 20 20:58 plugins

carrodher commented:
> @andresbono when running this chart with a securityContext it does not work.
> the plugins directory does not have the correct uid and gid set
> 
> ```
> 1001@5677fdebe960:/opt/bitnami/grafana/data$ ls -lah
> > total 12K
> > drwxrwxr-x  3 root root 4.0K Dec 20 20:57 .
> > drwxr-xr-x 11 1001 1001 4.0K Dec 20 20:58 ..
> > drwxrwxr-x  2 root root 4.0K Dec 20 20:58 plugins
> ```
> 
> the logs tell the same story
> 
> ```
>  >>> kubectl logs grafana-deployment-54c59bb487-mmzjg --namespace grafana-operator-8whwaptevm --container grafana
> Ensuring '/opt/bitnami/grafana/conf/grafana.ini' points to '/etc/grafana/grafana.ini'
> Ensuring '/opt/bitnami/grafana/data' points to '/var/lib/grafana'
> rm: cannot remove '/opt/bitnami/grafana/data/plugins': Permission denied
> ```

It is possible we need to add the `volumePermission` initcontainer as we're doing in other charts, for example:
- https://github.com/bitnami/charts/blob/d013f37e89f9e1111efcb334607a3aceb7569e1d/bitnami/redmine/values.yaml#L772
- https://github.com/bitnami/charts/blob/d013f37e89f9e1111efcb334607a3aceb7569e1d/bitnami/redmine/templates/deployment.yaml#L55

Cellebyte commented:
@carrodher will add the second approach to the `Grafana` instance and we will see how it goes ;)

Cellebyte commented:
@carrodher the operator does not allow adding a container.
The initContainer is started with the correct podSecurityContext.
I think its an issue with how the plugins are initialized with the grafana container.
The script runs as root and creates the folder as root at image build time.

Cellebyte commented:
I think this line
https://github.com/bitnami/bitnami-docker-grafana/blob/c38f2826132b8dfeb4621a189eda4790ee509ca2/7/debian-10/Dockerfile#L23

needs to move under the USER 1001 field.

Cellebyte commented:
@javsalgar who can fix the grafana image?
I proposed the change which is needed to make the container work with the operator.
I think it is the only piece missing.
After you introduce the change the tests should run through.

Cellebyte commented:
@1337andre @andresbono @carrodher @joancafom I think the only thing missing is the fix in the grafana image build.
I think after that we can start with final review 👍

Cellebyte commented:
@andresbono wanted to fix it right away but you were faster :D :+1:  Thanks :)

Cellebyte commented:
The PR ia now in a state where we can have a final review round and hopefully can merge it soon.
Thanks for all who participated in this PR :D
I can gladly say that this Chart is now running as non root which is awesome :)

@andresbono thanks for your time and support.

andresbono reviewed (approved):
LGTM

andresbono commented:
> The PR is now in a state where we can have a final review round and hopefully can merge it soon.
> Thanks for all who participated in this PR :D
> I can gladly say that this Chart is now running as non root which is awesome :)
> 
> @andresbono thanks for your time and support.

Thanks to you, @Cellebyte. You did 99% of the heavy lifting to get this chart ready. Another great contribution!!
